File: solutions/lua/luhn/1/luhn.lua
return {
	valid = function(s)
		s = string.gsub(s, "%s", "")
		if string.gsub(s, "%d", "") ~= "" then
			return false
		end
		s = string.gsub(s, "%D", "")
		local l = string.len(s)
		if l < 2 then
			return false
		end
		local sum, rev = 0, string.reverse(s)
		for i = 1, l, 2 do
			local odd = tonumber(string.sub(rev, i, i)) or 0
			local even = 2 * (tonumber(string.sub(rev, i + 1, i + 1)) or 0)
			if even > 9 then
				even = even - 9
			end
			sum = sum + even + odd
		end
		return sum % 10 == 0
	end,
}

File: solutions/ruby/luhn/1/luhn.rb
# frozen_string_literal: true

# Luhn class
class Luhn
  def self.valid?(str)
    digits = str.gsub(/\d/).to_a.map(&:to_i)
    return false if digits.size <= 1
    return false if str.gsub(/[^\d\s]/).to_a.size.positive?

    every1 = digits.select.with_index { _2.even? != digits.length.even? }
    every2 = digits.select.with_index { _2.even? == digits.length.even? }
    doubles = every2.lazy.map { _1 * 2 }.map { _1 > 9 ? _1 - 9 : _1 }

    ((doubles.sum + every1.sum) % 10).zero?
  end
end
